### PR TITLE
Authorization, Cleanup, and Logout

### PR DESCRIPTION
**Pull request by Mariamcm90**
This PR aims to address some of the shortcomings of the PR I made earlier in the week. 

Adds the use of JSON Web Tokens in order to protect routes on the front-end. Initially, this was done by making a request to an 'auth' route and determining if the response had a user. However, this was slow and caused the app to crash if you were not logged in and on the home page, because during this time the routes necessary to load the page would be called. I figure we'd eventually have to find a better way to protect routes, especially for private viewing party rooms so that only certain users can access certain routes. 

Adds logout capability. You can access it by clicking on the user profile button on the nav bar which will display a dropdown where you can log out. 

If a user attempts to access a route they are not authorized to, they get redirected back to the login page. If the user is logged in and tries to access the login page, it redirects them directly to the home page (if they want to access the page they must first log out).

Some caveats to this are: 
With JWTs most guides take one of two paths (storing in sessionStorage or in cookies). Both are inherently on their own insecure. With the former, you run the risk of cross-site scripting and with the latter cross-site forgery, and so usually if used it's used with HTTPS, input sanitization, etc which seem beyond the scope of this project. Not super relevant to the PR but interesting to learn about. 

Allows it so in the profile edit view, previous genres that the user has selected are pre-highlighted and the input box for a profile description is pre-populated. 

Breaks up reducers and action files. As these files get bigger, I thought it would be better to have them separated by concern. 

Cleans up some of the issues that were being logged in the development console to make it faster to find log statements that you insert for testing. 

I still haven't gotten around to notifying the user on the front-end if login/registration is unsuccessful but I've made it in the meantime so that this fails more gracefully and clearly logs the issue. I believe previously it would crash (something to do with the order of flash middleware not being in the right order relative to something else).

### CODE REVIEW

**Comment by matthewbinshtok** [comment]
> If a user attempts to access a route they are not authorized to, they get redirected back to the login page. If the user is logged in and tries to access the login page, it redirects them directly to the home page (if they want to access the page they must first log out).

Great revision for user flow! 👍 

> Breaks up reducers and action files. As these files get bigger, I thought it would be better to have them separated by concern.

Good call. Have this in my next PR, but I'll resolve the conflicts and use your structure.

File: client/src/actions/index.js [hunk 1]
```diff
@@ -1,178 +1,5 @@
...
-
-export const updateProfileGenres = (genres) => {
-    return {
-        type: PROFILE_GENRES_UPDATED,
-        payload: genres
-    };
-};
+export * from './contentActions';
+export * from './profileActions';
+export * from './authActions';
```
Comment (matthewbinshtok) [inline client/src/actions/index.js]: Great stuff.

File: client/src/components/ContentFeed/ContentFeed.js [hunk 2]
```diff
@@ -4,12 +4,12 @@ import {connect} from "react-redux";
 import './ContentFeed.css'
 import ContentTile from '../ContentTile/ContentTile';
 
-import {getMovieFeedContent, getShowFeedContent} from '../../actions/index';
+import {getMovieFeedContent, getShowFeedContent} from '../../actions/contentActions';
```
Comment (matthewbinshtok) [inline client/src/components/ContentFeed/ContentFeed.js]: Are we sure you need to change this if actions are getting exported from index with the revision above?

File: client/src/components/Home/Home.js [hunk 3]
```diff
@@ -4,16 +4,33 @@ import NavigationBar from '../NavigationBar/NavigationBar'
 import ContentFeed from '../ContentFeed/ContentFeed'
 import Welcome from '../Welcome/Welcome'
 import EditWelcome from '../EditWelcome/EditWelcome'
-import './Home.css';
-import axios from 'axios';
+import LoginRegistration from '../LoginRegistration/LoginRegistration'
 import {connect} from "react-redux";
-import {getProfileInformation} from "../../actions";
+import {getProfileInformation} from "../../actions/profileActions";
```
Comment (matthewbinshtok) [inline client/src/components/Home/Home.js]: Same comment about exports as above...

**Comment by matthewbinshtok** [comment]
Ok, left a few comments. If you address those, it's good to do.
As for importing action types, let me know what you think about this approach:
[](https://medium.com/@nate_wang/a-new-approach-for-managing-redux-actions-91c26ce8b5da)

Fantastic work! Get this bread! 💪

File: client/src/components/ContentFeed/ContentFeed.js (hunk 2 above)
Comment (Mariamcm90) [reply client/src/components/ContentFeed/ContentFeed.js]: Nope, not at all. Thanks for catching that!

File: client/src/components/ContentFeed/ContentFeed.js [hunk 4]
```diff
@@ -48,8 +48,8 @@ class ConnectedContentFeed extends Component {
         let contentTileList = [];
 
         if (recommendations !== null) {
-            contentTileList = recommendations.map((content_element) => (
-                <ContentTile metadata={content_element}/>
+            contentTileList = recommendations.map((content_element, i) => (
+                <ContentTile key={i} metadata={content_element}/>
```
Comment (matthewbinshtok) [inline client/src/components/ContentFeed/ContentFeed.js]: Thank you!!

File: client/src/components/EditWelcome/EditWelcome.js [hunk 5]
```diff
@@ -1,17 +1,17 @@
 import React, {Component} from "react";
 import {withRouter} from 'react-router-dom'
+import {connect} from "react-redux";
 import ProfilePicture from '../ProfilePicture/ProfilePicture'
 import './EditWelcome.css';
-
-import {connect} from "react-redux";
-import {updateProfileView, updateProfileInformation, getProfileInformation} from '../../actions/index';
+import {updateProfileView, updateProfileInformation} from '../../actions/profileActions';
```
Comment (matthewbinshtok) [inline client/src/components/EditWelcome/EditWelcome.js]: ../../actions

File: client/src/components/ProfilePicture/ProfilePicture.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ import {connect} from "react-redux";
 import {getProfilePicture, uploadProfilePicture} from '../../actions/index';
```
Comment (matthewbinshtok) [inline client/src/components/ProfilePicture/ProfilePicture.js]: ../../actions

File: client/src/components/SearchBar/SearchBar.js [hunk 7]
```diff
@@ -5,7 +5,7 @@ import { connect } from "react-redux";
 import { getResults } from '../../actions/index';
```
Comment (matthewbinshtok) [inline client/src/components/SearchBar/SearchBar.js]: ../../actions

File: client/src/components/Welcome/Welcome.js [hunk 8]
```diff
@@ -1,18 +1,19 @@
 import React, {Component} from "react";
 import {withRouter} from 'react-router-dom'
-import ProfilePicture from '../ProfilePicture/ProfilePicture'
 import { connect } from "react-redux";
+import {getProfileInformation, updateProfileView} from "../../actions/profileActions";
```
Comment (matthewbinshtok) [inline client/src/components/Welcome/Welcome.js]: ../../actions

**Pull request merged**